File: agent-dotnet/SecurityAgent/Formatters/BasicFormatter.cs
using SecurityAgent.Models;

namespace SecurityAgent.Formatters
{
    /// <summary>
    /// Basic log formatter (FIM, PROC, NET, AUTH, REG, SVC)
    /// Compatible with LogDecoder from the main system
    /// </summary>
    public class BasicFormatter : IFormatter
    {
        public string FormatFileIntegrity(FileChange change)
        {
            // Format: FIM [YYYY-MM-DD HH:MM:SS] ACTION: FILE_PATH
            var timestamp = change.Timestamp.ToString("yyyy-MM-dd HH:mm:ss");
            return $"FIM [{timestamp}] {change.Action}: {change.FilePath}";
        }

        public string FormatProcess(ProcessInfo process)
        {
            // Format: PROC [YYYY-MM-DD HH:MM:SS] PID:XXX User:USER Cmd:COMMAND
            var timestamp = process.CreateTime.ToString("yyyy-MM-dd HH:mm:ss");
            var command = !string.IsNullOrEmpty(process.CommandLine)
                ? process.CommandLine
                : process.ProcessName;

            return $"PROC [{timestamp}] PID:{process.ProcessId} User:{process.UserName} Cmd:{command}";
        }

        public string FormatNetwork(NetworkConnection connection)
        {
            // Format: NET [YYYY-MM-DD HH:MM:SS] PROTOCOL SRC_IP:SRC_PORT -> DST_IP:DST_PORT
            var timestamp = connection.Timestamp.ToString("yyyy-MM-dd HH:mm:ss");
            return $"NET [{timestamp}] {connection.Protocol} {connection.LocalAddress}:{connection.LocalPort} -> {connection.RemoteAddress}:{connection.RemotePort}";
        }

        public string FormatAuthentication(string result, string user, string source, DateTime timestamp)
        {
            // Format: AUTH [YYYY-MM-DD HH:MM:SS] RESULT User:USER From:SOURCE
            var time = timestamp.ToString("yyyy-MM-dd HH:mm:ss");
            return $"AUTH [{time}] {result} User:{user} From:{source}";
        }

        public string FormatRegistry(RegistryChange change)
        {
            // Format: REG [YYYY-MM-DD HH:MM:SS] ACTION: Key:KEY_PATH Value:VALUE
            var timestamp = change.Timestamp.ToString("yyyy-MM-dd HH:mm:ss");
            var value = !string.IsNullOrEmpty(change.ValueName)
                ? $"{change.ValueName}={change.ValueData}"
                : change.ValueData;

            return $"REG [{timestamp}] {change.Action}: Key:{change.KeyPath} Value:{value}";
        }

        public string FormatService(string action, string serviceName, string status, DateTime timestamp)
        {
            // Format: SVC [YYYY-MM-DD HH:MM:SS] ACTION: Service:SERVICE_NAME Status:STATUS
            var time = timestamp.ToString("yyyy-MM-dd HH:mm:ss");
            return $"SVC [{time}] {action}: Service:{serviceName} Status:{status}";
        }
    }
}

File: agent/formatters/basic_formatter.py
"""
Basic Log Formatter
Formats logs into FIM, PROC, NET, AUTH, REG, SVC formats
Compatible with LogDecoder from the main system
"""
from datetime import datetime
from typing import Dict, Any


class BasicFormatter:
    """
    Formats security events into basic log format compatible with LogDecoder.
    Supported formats: FIM, PROC, NET, AUTH, REG, SVC
    """

    @staticmethod
    def format_file_integrity(file_path: str, action: str, timestamp: datetime = None) -> str:
        """
        Format: FIM [YYYY-MM-DD HH:MM:SS] ACTION: FILE_PATH
        Example: FIM [2025-11-05 10:19:00] modified: C:\Windows\System32\drivers\etc\hosts
        """
        if timestamp is None:
            timestamp = datetime.now()

        timestamp_str = timestamp.strftime('%Y-%m-%d %H:%M:%S')
        return f"FIM [{timestamp_str}] {action}: {file_path}"

    @staticmethod
    def format_process(pid: int, user: str, command: str, timestamp: datetime = None) -> str:
        """
        Format: PROC [YYYY-MM-DD HH:MM:SS] PID:XXX User:USER Cmd:COMMAND
        Example: PROC [2025-11-05 10:20:00] PID:1234 User:admin Cmd:powershell.exe -enc XXX
        """
        if timestamp is None:
            timestamp = datetime.now()

        timestamp_str = timestamp.strftime('%Y-%m-%d %H:%M:%S')
        return f"PROC [{timestamp_str}] PID:{pid} User:{user} Cmd:{command}"

    @staticmethod
    def format_network(protocol: str, src_ip: str, src_port: int,
                       dst_ip: str, dst_port: int, timestamp: datetime = None) -> str:
        """
        Format: NET [YYYY-MM-DD HH:MM:SS] PROTOCOL SRC_IP:SRC_PORT -> DST_IP:DST_PORT
        Example: NET [2025-11-05 10:21:00] TCP 192.168.1.10:5000 -> 8.8.8.8:443
        """
        if timestamp is None:
            timestamp = datetime.now()

        timestamp_str = timestamp.strftime('%Y-%m-%d %H:%M:%S')
        return f"NET [{timestamp_str}] {protocol} {src_ip}:{src_port} -> {dst_ip}:{dst_port}"

    @staticmethod
    def format_authentication(result: str, user: str, source: str, timestamp: datetime = None) -> str:
        """
        Format: AUTH [YYYY-MM-DD HH:MM:SS] RESULT User:USER From:SOURCE
        Example: AUTH [2025-11-05 10:22:00] failed User:admin From:192.168.1.100
        """
        if timestamp is None:
            timestamp = datetime.now()

        timestamp_str = timestamp.strftime('%Y-%m-%d %H:%M:%S')
        return f"AUTH [{timestamp_str}] {result} User:{user} From:{source}"

    @staticmethod
    def format_registry(action: str, key: str, value: str, timestamp: datetime = None) -> str:
        """
        Format: REG [YYYY-MM-DD HH:MM:SS] ACTION: Key:KEY_PATH Value:VALUE
        Example: REG [2025-11-05 10:23:00] modified: Key:HKLM\Software\Run Value:malware.exe
        """
        if timestamp is None:
            timestamp = datetime.now()

        timestamp_str = timestamp.strftime('%Y-%m-%d %H:%M:%S')
        return f"REG [{timestamp_str}] {action}: Key:{key} Value:{value}"

    @staticmethod
    def format_service(action: str, service_name: str, status: str, timestamp: datetime = None) -> str:
        """
        Format: SVC [YYYY-MM-DD HH:MM:SS] ACTION: Service:SERVICE_NAME Status:STATUS
        Example: SVC [2025-11-05 10:24:00] stopped: Service:WinDefend Status:stopped
        """
        if timestamp is None:
            timestamp = datetime.now()

        timestamp_str = timestamp.strftime('%Y-%m-%d %H:%M:%S')
        return f"SVC [{timestamp_str}] {action}: Service:{service_name} Status:{status}"
